### DISCUSSION TITLE
Protecting the investment in Vulnerability Analysis

### DISCUSSION THREAD

**Discussion by flemminglau** [DependencyTrack/dependency-track#1961] [Ideas]
One great thing about DT is that basically all data in there can be automatically recreated from the BOM files and automatically from external sources. 

Except from the Vulnerability handling. This will be entered manually and as it looks at the moment cannot easily be exported and restored. 

(I just raised https://github.com/DependencyTrack/dependency-track/issues/1958 so it is not only difficult. Actually it is impossible to restore as it stands.)

If using DT for real for this the Violation handling can represent countless hours of invested analysis work. It would be really great if this information could be safeguarded by retaining it with the source code and not only in an external system which may disappear some day. 

Is this aspect being looked at?

One option would be to support the exporting of the data in a format which could also be loaded. Or at least support that any "False Positive"->"Suppress" information can be loaded from an external source. Currently it seems one via the API has to identify such information via the Vulnerability UUID which is not very portable.

**Comment by syalioune** [comment]
Hello @flemminglau 

DT support [VEX](https://cyclonedx.org/capabilities/vex/) which fit right into your use case from where I stand. A VEX is basically a SBOM with a `vulnerabilities` node conveying audit state that you can import/export through UI or API 

![image](https://user-images.githubusercontent.com/6144741/190398131-f41bf28f-8012-4d68-b598-5a0bb7d339fc.png)
![image](https://user-images.githubusercontent.com/6144741/190398509-15439ce4-3919-4cd0-a858-5efd44d1195f.png)

That's a least the way we are dealing with this in my company.

You can find some examples here https://github.com/CycloneDX/bom-examples/tree/master/VEX

**Reply by flemminglau** [reply]
In case a VEX file is effectively a complete backup of a Project including the BOM information, any automatically generated vulnerability information and all info entered into the Vulnerability Audit fields you are right.

It seems however the Suppress field is not included?

In case it is indeed a complete copy we can export the VEX files and store them with our sources in case we want to restore the entered information somewhere else. I assume even into a non-DT system that supports loading VEX files.

**Reply by nscuro** [reply]
VEX doesn't have a concept of explicit suppressions. Instead suppressions can be inferred based on the analysis state.

DT will treat analyses with state `FALSE_POSITIVE`, `NOT_AFFECTED` and `RESOLVED` as suppressions when importing VEX. 

And yes, VEX is a universal concept. I recommend you look into some of the official guidance by CISA, in addition to the CycloneDX link @syalioune shared earlier:

* https://www.cisa.gov/sites/default/files/publications/VEX_Use_Cases_Aprill2022.pdf
* https://www.cisa.gov/sites/default/files/publications/VEX_Status_Justification_Jun22.pdf

**Reply by nscuro** [reply]
If VEX doesn't satisfy your needs, you can still use DT's findings API to create exports: https://docs.dependencytrack.org/integrations/file-formats/

This however is DT-specific, and there is not support for importing that data again (but it can be trivially scripted if you need it).

Also note that neither VEX nor the findings API cover policy violations.

**Reply by flemminglau** [reply]
Just for completeness:
What happens if I was to load a VEX file onto an existing already scanned and analyzed project?
What happens to existing vulnerability handling information?
Specifically what happens if someone had set Suppress without setting the Analysis to RESOLVED or similar?
What happens at the next scan for vulnerabilities?

**Reply by syalioune** [reply]
1. DT VEX importer perform a delta update. Let's say, you had component A (no analysis), B (false_positive). If you upload a VEX with A (exploitable), you'll end up with A (exploitable) and B (false_positive)
2. Same answer, they are kept
3. Only test will tell, I personally don't have the answer 😄

**Reply by nscuro** [reply]
> Specifically what happens if someone had set Suppress without setting the Analysis to RESOLVED or similar?

That suppression will be lost. Because as stated before, VEX doesn't cover explicit suppressions.

> What happens at the next scan for vulnerabilities?

Nothing special, really. Applying a VEX is not that much different from doing it manually. From the import onwards it "behaves" the same.

**Reply by flemminglau** [reply]
Ok. 
So one should see the loading of a VEX file as an automated editing of the existing data.
With the tweak of the generation of the Suppress flag:
- Any existing data mentioned in the VEX will update existing data
- Any new data in the VEX will create new data
- There is no way in the VEX to remove existing data

Right?

**Comment by syalioune** [comment]
That's my understanding.
